diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
index 4698933a30ebc..68164f6e74fc8 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
@@ -16,14 +16,17 @@
 #include <linux/module.h>
 #include <linux/mutex.h>
 #include <linux/of.h>
+#include <linux/workqueue.h>
 
 #include <drm/bridge/dw_hdmi_qp.h>
 #include <drm/display/drm_hdmi_helper.h>
 #include <drm/display/drm_hdmi_cec_helper.h>
 #include <drm/display/drm_hdmi_state_helper.h>
+#include <drm/display/drm_scdc_helper.h>
 #include <drm/drm_atomic.h>
 #include <drm/drm_atomic_helper.h>
 #include <drm/drm_bridge.h>
+#include <drm/drm_bridge_helper.h>
 #include <drm/drm_connector.h>
 #include <drm/drm_edid.h>
 #include <drm/drm_modes.h>
@@ -41,6 +44,22 @@
 
 #define HDMI14_MAX_TMDSCLK	340000000
 
+/*
+ * Recommended N and Expected CTS Values in FRL Mode.
+ */
+static const struct dw_hdmi_qp_audio_frl_n {
+	unsigned int r_bit;
+	unsigned int n_32k;
+	unsigned int n_44k1;
+	unsigned int n_48k;
+} common_frl_n_table[] = {
+	{ .r_bit = 3,  .n_32k = 4224, .n_44k1 = 5292, .n_48k = 5760, },
+	{ .r_bit = 6,  .n_32k = 4032, .n_44k1 = 5292, .n_48k = 6048, },
+	{ .r_bit = 8,  .n_32k = 4032, .n_44k1 = 3969, .n_48k = 6048, },
+	{ .r_bit = 10, .n_32k = 3456, .n_44k1 = 3969, .n_48k = 5184, },
+	{ .r_bit = 12, .n_32k = 3072, .n_44k1 = 3969, .n_48k = 4752, },
+};
+
 #define SCRAMB_POLL_DELAY_MS	3000
 
 struct dw_hdmi_qp_i2c {
@@ -88,6 +107,8 @@ struct dw_hdmi_qp {
 
 	/* Written by the atomic enable/disable hooks, read locklessly by HPD */
 	struct drm_connector *curr_conn;
+	struct work_struct flt_work;
+	bool flt_no_timeout;
 	unsigned long tmds_char_rate;
 	bool no_hpd;
 };
@@ -136,6 +157,49 @@ static void dw_hdmi_qp_set_cts_n(struct dw_hdmi_qp *hdmi, unsigned int cts,
 		       AUDPKT_ACR_CONTROL1);
 }
 
+static int dw_hdmi_qp_match_frl_n_table(struct dw_hdmi_qp *hdmi,
+					unsigned long r_bit,
+					unsigned long freq)
+{
+	const struct dw_hdmi_qp_audio_frl_n *frl_n = NULL;
+	int i, n;
+
+	for (i = 0; i < ARRAY_SIZE(common_frl_n_table); i++) {
+		if (r_bit == common_frl_n_table[i].r_bit) {
+			frl_n = &common_frl_n_table[i];
+			break;
+		}
+	}
+
+	if (!frl_n) {
+		dev_err(hdmi->dev, "Unexpected FRL Rbit: %lu Gbps\n", r_bit);
+		return 0;
+	}
+
+	switch (freq) {
+	case 32000:
+	case 64000:
+	case 128000:
+		n = (freq / 32000) * frl_n->n_32k;
+		break;
+	case 44100:
+	case 88200:
+	case 176400:
+		n = (freq / 44100) * frl_n->n_44k1;
+		break;
+	case 48000:
+	case 96000:
+	case 192000:
+		n = (freq / 48000) * frl_n->n_48k;
+		break;
+	default:
+		dev_err(hdmi->dev, "Unexpected FRL freq: %lu Hz\n", freq);
+		n = 0;
+	}
+
+	return n;
+}
+
 static void dw_hdmi_qp_set_audio_interface(struct dw_hdmi_qp *hdmi,
 					   struct hdmi_codec_daifmt *fmt,
 					   struct hdmi_codec_params *hparms)
@@ -253,9 +317,18 @@ static void dw_hdmi_qp_set_channel_status(struct dw_hdmi_qp *hdmi,
 static void dw_hdmi_qp_set_sample_rate(struct dw_hdmi_qp *hdmi, unsigned long long tmds_char_rate,
 				       unsigned int sample_rate)
 {
+	const struct dw_hdmi_qp_link_cfg *link_cfg;
 	unsigned int n, cts;
 
-	drm_hdmi_acr_get_n_cts(tmds_char_rate, sample_rate, &n, &cts);
+	link_cfg = hdmi->phy.ops->get_link_cfg(hdmi, hdmi->phy.data);
+	if (link_cfg->frl_enabled) {
+		n = dw_hdmi_qp_match_frl_n_table(hdmi,
+						 link_cfg->frl_rate_per_lane,
+						 sample_rate);
+		cts = 0;
+	} else {
+		drm_hdmi_acr_get_n_cts(tmds_char_rate, sample_rate, &n, &cts);
+	}
 
 	dw_hdmi_qp_set_cts_n(hdmi, cts, n);
 }
@@ -547,12 +620,476 @@ static struct i2c_adapter *dw_hdmi_qp_i2c_adapter(struct dw_hdmi_qp *hdmi)
 	return adap;
 }
 
+enum dw_hdmi_qp_frl_lts {
+	LTS1,	/* Read EDID */
+	LTS2,	/* Prepare for FRL */
+	LTS3,	/* Training in progress */
+	LTS4,	/* Update FRL rate */
+	LTSP,	/* Training passed */
+	LTSL,	/* Legacy TMDS */
+	LTSU,	/* Undefined */
+};
+
+/*
+ * Check sink version and FLT no-timeout mode.
+ */
+static int dw_hdmi_qp_frl_lts1(struct dw_hdmi_qp *hdmi)
+{
+	int ret;
+	u8 val;
+
+	if (!hdmi->tmds_char_rate) {
+		dev_dbg(hdmi->dev, "lts1: hdmi disabled\n");
+		return LTSL;
+	}
+
+	dw_hdmi_qp_mod(hdmi, AVP_DATAPATH_VIDEO_SWDISABLE,
+		       AVP_DATAPATH_VIDEO_SWDISABLE, GLOBAL_SWDISABLE);
+
+	/* Reset AVP data path */
+	dw_hdmi_qp_write(hdmi, AVP_DATAPATH_SWINIT_P, GLOBAL_SWRESET_REQUEST);
+
+	ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_SINK_VERSION, &val);
+	if (ret) {
+		dev_err(hdmi->dev, "lts1: SCDC read failed\n");
+		return LTSL;
+	}
+
+	if (!val) {
+		dev_warn(hdmi->dev, "lts1: SCDC sink version is zero\n");
+		return LTSL;
+	}
+
+	ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_SOURCE_VERSION, 1);
+	if (ret) {
+		dev_err(hdmi->dev, "lts1: SCDC write failed\n");
+		return LTSL;
+	}
+
+	ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_SOURCE_TEST_CONFIG, &val);
+	if (ret) {
+		dev_err(hdmi->dev, "lts1: SCDC read failed\n");
+		return LTSL;
+	}
+
+	hdmi->flt_no_timeout = !!(val & SCDC_FLT_NO_TIMEOUT);
+	dev_dbg(hdmi->dev, "lts1: flt_no_timeout=%d\n", hdmi->flt_no_timeout);
+
+	return LTS2;
+}
+
+/*
+ * Check if sink is ready to training. Set FRL rate & max FFE level.
+ */
+static int dw_hdmi_qp_frl_lts2(struct dw_hdmi_qp *hdmi)
+{
+	const struct dw_hdmi_qp_link_cfg *link_cfg;
+	int ret, i;
+	u8 val;
+
+	for (i = 0; i < 20; i++) {
+		ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_STATUS_FLAGS_0, &val);
+		if (ret) {
+			dev_err(hdmi->dev, "lts2: SCDC read failed\n");
+			return LTSL;
+		}
+
+		if (val & SCDC_FLT_READY) {
+			link_cfg = hdmi->phy.ops->get_link_cfg(hdmi, hdmi->phy.data);
+			val = hdmi->phy.ops->set_ffe_level ? link_cfg->max_ffe_level : 0;
+
+			dev_dbg(hdmi->dev, "lts2: set rate=%ux%u maxffe=%u\n",
+				link_cfg->frl_rate_per_lane, link_cfg->frl_lanes, val);
+
+			ret = drm_scdc_set_frl(hdmi->curr_conn, link_cfg->frl_rate_per_lane,
+					       link_cfg->frl_lanes, val);
+			if (ret)
+				ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_CONFIG_0, 0);
+			else
+				ret = -EIO;
+
+			if (ret) {
+				dev_err(hdmi->dev, "lts2: SCDC write failed\n");
+				return LTSL;
+			}
+
+			return LTS3;
+		}
+
+		msleep(20);
+	}
+
+	dev_err(hdmi->dev, "lts2: sink flt not ready\n");
+	return LTSL;
+}
+
+/*
+ * Conduct link training for the specified FRL rate.
+ */
+static int dw_hdmi_qp_frl_lts3(struct dw_hdmi_qp *hdmi)
+{
+	u8 val, ffe_lv = 0;
+	int i, ret;
+
+	/* Set 2s timeout */
+	i = 4000;
+
+	while (i-- > 0 || hdmi->flt_no_timeout) {
+		/* Poll FLT_update flag every 2 ms or less */
+		usleep_range(400, 500);
+
+		if (!hdmi->tmds_char_rate) {
+			dev_dbg(hdmi->dev, "lts3: hdmi disabled\n");
+			return LTSL;
+		}
+
+		ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_UPDATE_0, &val);
+		if (ret) {
+			dev_err(hdmi->dev, "lts3: SCDC read failed\n");
+			return LTSL;
+		}
+
+		if (val & SCDC_SOURCE_TEST_UPDATE) {
+			u8 test_cfg;
+
+			ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_SOURCE_TEST_CONFIG,
+					     &test_cfg);
+			if (ret) {
+				dev_err(hdmi->dev, "lts3: SCDC read failed\n");
+				return LTSL;
+			}
+
+			if (hdmi->flt_no_timeout && !(test_cfg & SCDC_FLT_NO_TIMEOUT)) {
+				dev_dbg(hdmi->dev, "lts3: exit test mode\n");
+				hdmi->flt_no_timeout = false;
+			} else if (!hdmi->flt_no_timeout && (test_cfg & SCDC_FLT_NO_TIMEOUT)) {
+				dev_dbg(hdmi->dev, "lts3: enter test mode\n");
+				hdmi->flt_no_timeout = true;
+			}
+
+			/* Clear SCDC_SOURCE_TEST_UPDATE flag */
+			ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0,
+					      SCDC_SOURCE_TEST_UPDATE);
+			if (ret) {
+				dev_err(hdmi->dev, "lts3: SCDC write failed\n");
+				return LTSL;
+			}
+		}
+
+		if (val & SCDC_FLT_UPDATE) {
+			u8 ln0, ln1, ln2, ln3;
+			u32 flt_cfg;
+
+			ret = drm_scdc_get_frl_ltp_request(hdmi->curr_conn,
+							   &ln0, &ln1, &ln2, &ln3);
+			if (!ret) {
+				dev_err(hdmi->dev, "lts3: SCDC read failed\n");
+				return LTSL;
+			}
+
+			dev_dbg(hdmi->dev, "lts3: ln0=0x%x ln1=0x%x ln2=0x%x ln3=0x%x\n",
+				ln0, ln1, ln2, ln3);
+
+			if (!ln0 && !ln1 && !ln2 && !ln3) {
+				dw_hdmi_qp_write(hdmi, 0, FLT_CONFIG1);
+				return LTSP;
+			}
+
+			if (ln0 == 0xf && ln1 == 0xf && ln2 == 0xf && ln3 == 0xf) {
+				dev_dbg(hdmi->dev, "lts3: rate change request\n");
+				return LTS4;
+			}
+
+			if (ln0 == 0xe || ln1 == 0xe || ln2 == 0xe || ln3 == 0xe) {
+				if (!hdmi->phy.ops->set_ffe_level) {
+					dev_err(hdmi->dev, "lts3: ffe level update not expected\n");
+					return LTSL;
+				}
+
+				if (ffe_lv >= 3) {
+					dev_err(hdmi->dev, "lts3: ffe level limit reached\n");
+					return LTSL;
+				}
+
+				ffe_lv++;
+				dev_dbg(hdmi->dev, "lts3: ffe level up %d\n", ffe_lv);
+
+				ret = hdmi->phy.ops->set_ffe_level(hdmi, hdmi->phy.data, ffe_lv);
+				if (ret)
+					return LTSL;
+			} else {
+				flt_cfg = (ln3 << 16) | (ln2 << 12) | (ln1 << 8) |
+					  (ln0 << 4) | 0xf;
+
+				/* Support HFR1-10; send old LTP if ln0..3 == 0x3 */
+				if (!hdmi->flt_no_timeout && flt_cfg == 0x3333f)
+					flt_cfg = dw_hdmi_qp_read(hdmi, FLT_CONFIG1);
+
+				dw_hdmi_qp_write(hdmi, flt_cfg, FLT_CONFIG1);
+			}
+
+			/* Clear FLT_update flag */
+			ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0,
+					      SCDC_FLT_UPDATE);
+			if (ret) {
+				dev_err(hdmi->dev, "lts3: SCDC write failed\n");
+				return LTSL;
+			}
+		}
+	}
+
+	dev_err(hdmi->dev, "lts3: timed out\n");
+	return LTSL;
+}
+
+/*
+ * Handle FRL rate change request from sink.
+ */
+static int dw_hdmi_qp_frl_lts4(struct dw_hdmi_qp *hdmi)
+{
+	const struct dw_hdmi_qp_link_cfg *link_cfg;
+	u8 try_rate_per_lane, try_lanes, max_ffe_level;
+	int ret;
+
+	link_cfg = hdmi->phy.ops->get_link_cfg(hdmi, hdmi->phy.data);
+
+	/* Choose a reduced bandwidth FRL rate */
+	ret = drm_scdc_calc_lower_frl(link_cfg->frl_rate_per_lane, link_cfg->frl_lanes,
+				      &try_rate_per_lane, &try_lanes);
+	if (!ret) {
+		dev_err(hdmi->dev, "lts4: failed to compute lower frl\n");
+		return LTSL;
+	}
+
+	if (try_rate_per_lane < link_cfg->min_frl_rate_per_lane ||
+	    try_lanes < link_cfg->min_frl_lanes) {
+		dev_err(hdmi->dev, "lts4: unsupported %ux%u rate\n",
+			try_rate_per_lane, try_lanes);
+		return LTSL;
+	}
+
+	dev_dbg(hdmi->dev, "lts4: switching to %ux%u\n", try_rate_per_lane, try_lanes);
+
+	ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0, SCDC_FLT_UPDATE);
+	if (ret) {
+		dev_err(hdmi->dev, "lts4: SCDC write failed\n");
+		return LTSL;
+	}
+
+	/* Disable phy */
+	hdmi->phy.ops->disable(hdmi, hdmi->phy.data);
+
+	ret = hdmi->phy.ops->set_frl_rate(hdmi, hdmi->phy.data,
+					  try_rate_per_lane, try_lanes);
+	if (ret) {
+		dev_err(hdmi->dev, "lts4: failed to set phy link rate\n");
+		return LTSL;
+	}
+
+	/* Enable phy */
+	hdmi->phy.ops->init(hdmi, hdmi->phy.data);
+
+	max_ffe_level = hdmi->phy.ops->set_ffe_level ? link_cfg->max_ffe_level : 0;
+
+	/* Set new rate */
+	ret = drm_scdc_set_frl(hdmi->curr_conn, try_rate_per_lane, try_lanes,
+			       max_ffe_level);
+	if (ret)
+		ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0,
+				      SCDC_FLT_UPDATE);
+	else
+		ret = -EIO;
+
+	if (ret) {
+		dev_err(hdmi->dev, "lts4: SCDC write failed\n");
+		return LTSL;
+	}
+
+	return LTS3;
+}
+
+/*
+ * Training passed, wait for further changes.
+ */
+static int dw_hdmi_qp_frl_ltsp(struct dw_hdmi_qp *hdmi)
+{
+	int i, ret;
+	u8 val;
+
+	ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0, SCDC_FLT_UPDATE);
+	if (ret) {
+		dev_err(hdmi->dev, "ltsp: SCDC write failed\n");
+		return LTSL;
+	}
+
+	/* Set 2s timeout */
+	i = 4000;
+
+	while (i--) {
+		/* Poll Update Flags every 2 ms or less */
+		usleep_range(400, 500);
+
+		if (!hdmi->tmds_char_rate) {
+			dev_dbg(hdmi->dev, "ltsp: hdmi disabled\n");
+			return LTSL;
+		}
+
+		ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_UPDATE_0, &val);
+		if (ret) {
+			dev_err(hdmi->dev, "ltsp: SCDC read failed\n");
+			return LTSL;
+		}
+
+		if (val & SCDC_FRL_START) {
+			dw_hdmi_qp_mod(hdmi, 0, AVP_DATAPATH_VIDEO_SWDISABLE,
+				       GLOBAL_SWDISABLE);
+
+			ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0,
+					      SCDC_FRL_START);
+			if (ret) {
+				dev_err(hdmi->dev, "ltsp: SCDC write failed\n");
+				return LTSL;
+			}
+
+			dw_hdmi_qp_write(hdmi, PKTSCHED_GCP_CLEAR_AVMUTE,
+					 PKTSCHED_PKT_CONTROL0);
+			dw_hdmi_qp_mod(hdmi, PKTSCHED_GCP_TX_EN, PKTSCHED_GCP_TX_EN,
+				       PKTSCHED_PKT_EN);
+
+			dev_dbg(hdmi->dev, "ltsp: flt success\n");
+			break;
+		}
+
+		if (val & SCDC_FLT_UPDATE) {
+			dw_hdmi_qp_mod(hdmi, AVP_DATAPATH_VIDEO_SWDISABLE,
+				       AVP_DATAPATH_VIDEO_SWDISABLE, GLOBAL_SWDISABLE);
+
+			ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0,
+					      SCDC_FLT_UPDATE);
+			if (ret) {
+				dev_err(hdmi->dev, "ltsp: SCDC write failed\n");
+				return LTSL;
+			}
+
+			return LTS3;
+		}
+	}
+
+	if (i < 0) {
+		dev_err(hdmi->dev, "ltsp: timed out\n");
+		return LTSL;
+	}
+
+	/* Ensure FLT_update flag is polled at least once every 250 ms. */
+	i = 5;
+
+	while (true) {
+		msleep(20);
+
+		if (!hdmi->tmds_char_rate) {
+			dev_dbg(hdmi->dev, "ltsp: hdmi disabled\n");
+			break;
+		}
+
+		if (i) {
+			i--;
+			continue;
+		}
+
+		ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_UPDATE_0, &val);
+		if (ret) {
+			dev_err(hdmi->dev, "ltsp: SCDC read failed\n");
+			break;
+		}
+
+		if (val & SCDC_FLT_UPDATE) {
+			dw_hdmi_qp_write(hdmi, PKTSCHED_GCP_SET_AVMUTE,
+					 PKTSCHED_PKT_CONTROL0);
+			dw_hdmi_qp_mod(hdmi, PKTSCHED_GCP_TX_EN, PKTSCHED_GCP_TX_EN,
+				       PKTSCHED_PKT_EN);
+
+			msleep(50);
+			dw_hdmi_qp_mod(hdmi, AVP_DATAPATH_VIDEO_SWDISABLE,
+				       AVP_DATAPATH_VIDEO_SWDISABLE, GLOBAL_SWDISABLE);
+
+			ret = drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0,
+					      SCDC_FLT_UPDATE);
+			if (ret) {
+				dev_err(hdmi->dev, "ltsp: SCDC write failed\n");
+				break;
+			}
+
+			return LTS2;
+		}
+
+		i = 5;
+	}
+
+	return LTSL;
+}
+
+/*
+ * Exit frl mode, i.e. for training failures or hdmi disabled.
+ */
+static int dw_hdmi_qp_frl_ltsl(struct dw_hdmi_qp *hdmi)
+{
+	enum drm_connector_status status;
+
+	status = hdmi->phy.ops->read_hpd(hdmi, hdmi->phy.data);
+
+	dev_dbg(hdmi->dev, "ltsl: conn_stat=%d\n", status);
+
+	if (status != connector_status_disconnected) {
+		drm_scdc_set_frl(hdmi->curr_conn, 0, 0, 0);
+		drm_scdc_writeb(hdmi->bridge.ddc, SCDC_UPDATE_0, SCDC_FLT_UPDATE);
+	}
+
+	dw_hdmi_qp_mod(hdmi, 0, AVP_DATAPATH_VIDEO_SWDISABLE, GLOBAL_SWDISABLE);
+
+	return LTSU;
+}
+
+static void dw_hdmi_qp_flt_work(struct work_struct *work)
+{
+	struct dw_hdmi_qp *hdmi = container_of(work, struct dw_hdmi_qp, flt_work);
+	enum dw_hdmi_qp_frl_lts state = LTS1;
+
+	while (true) {
+		switch (state) {
+		case LTS1:
+			state = dw_hdmi_qp_frl_lts1(hdmi);
+			break;
+		case LTS2:
+			state = dw_hdmi_qp_frl_lts2(hdmi);
+			break;
+		case LTS3:
+			state = dw_hdmi_qp_frl_lts3(hdmi);
+			break;
+		case LTS4:
+			state = dw_hdmi_qp_frl_lts4(hdmi);
+			break;
+		case LTSP:
+			state = dw_hdmi_qp_frl_ltsp(hdmi);
+			break;
+		case LTSL:
+			state = dw_hdmi_qp_frl_ltsl(hdmi);
+			break;
+		case LTSU:
+			return;
+		default:
+			dev_err(hdmi->dev, "unexpected flt state: %d\n", state);
+			return;
+		}
+	}
+}
+
 static void dw_hdmi_qp_bridge_atomic_enable(struct drm_bridge *bridge,
 					    struct drm_atomic_commit *state)
 {
 	struct dw_hdmi_qp *hdmi = bridge->driver_private;
 	struct drm_connector_state *conn_state;
 	struct drm_connector *connector;
+	const struct dw_hdmi_qp_link_cfg *link_cfg;
 	const struct drm_display_mode *mode;
 	struct drm_crtc_state *crtc_state;
 	unsigned int op_mode;
@@ -566,15 +1103,24 @@ static void dw_hdmi_qp_bridge_atomic_enable(struct drm_bridge *bridge,
 	if (WARN_ON(!conn_state))
 		return;
 
+	link_cfg = hdmi->phy.ops->get_link_cfg(hdmi, hdmi->phy.data);
+
 	if (connector->display_info.is_hdmi) {
 		crtc_state = drm_atomic_get_new_crtc_state(state, conn_state->crtc);
 		mode = &crtc_state->mode;
+		dev_dbg(hdmi->dev, "%s mode=%ux%u@%uHz fmt=%s rate=%llu bpc=%u\n", __func__,
+			mode->hdisplay, mode->vdisplay, drm_mode_vrefresh(mode),
+			drm_hdmi_connector_get_output_format_name(conn_state->hdmi.output_format),
+			conn_state->hdmi.tmds_char_rate, conn_state->hdmi.output_bpc);
+
 		op_mode = 0;
 		hdmi->tmds_char_rate = conn_state->hdmi.tmds_char_rate;
 
-		ret = drm_connector_hdmi_enable_scrambling(connector, conn_state);
-		if (ret)
-			dev_warn(hdmi->dev, "Failed to enable scrambling: %d\n", ret);
+		if (!link_cfg->frl_enabled) {
+			ret = drm_connector_hdmi_enable_scrambling(connector, conn_state);
+			if (ret)
+				dev_warn(hdmi->dev, "Failed to enable scrambling: %d\n", ret);
+		}
 
 		dev_dbg(hdmi->dev,
 			"%s mode=HDMI:%ux%u@%uHz fmt=%s rate=%llu bpc=%u scramb=%d\n",
@@ -593,9 +1139,23 @@ static void dw_hdmi_qp_bridge_atomic_enable(struct drm_bridge *bridge,
 	dw_hdmi_qp_mod(hdmi, HDCP2_BYPASS, HDCP2_BYPASS, HDCP2LOGIC_CONFIG0);
 	dw_hdmi_qp_mod(hdmi, op_mode, OPMODE_DVI, LINK_CONFIG0);
 
+	dw_hdmi_qp_mod(hdmi,
+		       link_cfg->frl_enabled && link_cfg->frl_lanes == 4 ? OPMODE_FRL_4LANES : 0,
+		       OPMODE_FRL_4LANES, LINK_CONFIG0);
+	dw_hdmi_qp_mod(hdmi, link_cfg->frl_enabled, OPMODE_FRL, LINK_CONFIG0);
+
 	drm_atomic_helper_connector_hdmi_update_infoframes(connector, state);
 
 	WRITE_ONCE(hdmi->curr_conn, connector);
+
+	if (link_cfg->frl_enabled) {
+		/* Ensure phy output is stable before starting FLT */
+		msleep(50);
+		schedule_work(&hdmi->flt_work);
+	} else {
+		dw_hdmi_qp_write(hdmi, PKTSCHED_GCP_CLEAR_AVMUTE, PKTSCHED_PKT_CONTROL0);
+		dw_hdmi_qp_mod(hdmi, PKTSCHED_GCP_TX_EN, PKTSCHED_GCP_TX_EN, PKTSCHED_PKT_EN);
+	}
 }
 
 static void dw_hdmi_qp_bridge_atomic_disable(struct drm_bridge *bridge,
@@ -604,20 +1164,62 @@ static void dw_hdmi_qp_bridge_atomic_disable(struct drm_bridge *bridge,
 	struct dw_hdmi_qp *hdmi = bridge->driver_private;
 	struct drm_connector *connector;
 
-	WRITE_ONCE(hdmi->curr_conn, NULL);
 	hdmi->tmds_char_rate = 0;
 
 	connector = drm_atomic_get_old_connector_for_encoder(state, bridge->encoder);
+
+	dw_hdmi_qp_write(hdmi, PKTSCHED_GCP_SET_AVMUTE, PKTSCHED_PKT_CONTROL0);
+	msleep(50);
+
+	cancel_work_sync(&hdmi->flt_work);
 	drm_connector_hdmi_disable_scrambling(connector);
+	WRITE_ONCE(hdmi->curr_conn, NULL);
 
 	hdmi->phy.ops->disable(hdmi, hdmi->phy.data);
 }
 
-static enum drm_connector_status
-dw_hdmi_qp_bridge_detect(struct drm_bridge *bridge, struct drm_connector *connector)
+static int dw_hdmi_qp_reset_crtc(struct dw_hdmi_qp *hdmi,
+				 struct drm_connector *connector,
+				 struct drm_modeset_acquire_ctx *ctx)
+{
+	u8 config;
+	int ret;
+
+	if (connector->hdmi.scrambler_enabled) {
+		ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_TMDS_CONFIG, &config);
+		if (ret < 0) {
+			dev_err(hdmi->dev, "Failed to read TMDS config: %d\n", ret);
+			return ret;
+		}
+
+		if (config & SCDC_SCRAMBLING_ENABLE)
+			return 0;
+	} else {
+		ret = drm_scdc_readb(hdmi->bridge.ddc, SCDC_CONFIG_1, &config);
+		if (ret < 0) {
+			dev_err(hdmi->dev, "Failed to read FRL config: %d\n", ret);
+			return ret;
+		}
+
+		if (config & SCDC_FRL_RATE_MASK && work_busy(&hdmi->flt_work))
+			return 0;
+	}
+
+	dev_dbg(hdmi->dev, "resetting crtc\n");
+
+	return drm_bridge_helper_reset_crtc(&hdmi->bridge, ctx);
+}
+
+static int dw_hdmi_qp_bridge_detect_ctx(struct drm_bridge *bridge,
+					struct drm_connector *connector,
+					struct drm_modeset_acquire_ctx *ctx)
 {
 	struct dw_hdmi_qp *hdmi = bridge->driver_private;
+	const struct dw_hdmi_qp_link_cfg *link_cfg;
+	enum drm_connector_status status;
 	const struct drm_edid *drm_edid;
+	bool frl_active;
+	int ret;
 
 	if (hdmi->no_hpd) {
 		drm_edid = drm_edid_read_ddc(connector, bridge->ddc);
@@ -627,7 +1229,24 @@ dw_hdmi_qp_bridge_detect(struct drm_bridge *bridge, struct drm_connector *connec
 			return connector_status_disconnected;
 	}
 
-	return hdmi->phy.ops->read_hpd(hdmi, hdmi->phy.data);
+	status = hdmi->phy.ops->read_hpd(hdmi, hdmi->phy.data);
+
+	link_cfg = hdmi->phy.ops->get_link_cfg(hdmi, hdmi->phy.data);
+	frl_active = link_cfg->frl_enabled && hdmi->tmds_char_rate;
+
+	dev_dbg(hdmi->dev, "%s status=%d frl=%u scramb=%u\n", __func__,
+		status, frl_active, connector->hdmi.scrambler_enabled);
+
+	if (status == connector_status_connected &&
+	    (frl_active || connector->hdmi.scrambler_enabled)) {
+		ret = dw_hdmi_qp_reset_crtc(hdmi, connector, ctx);
+		if (ret == -EDEADLK)
+			return ret;
+		if (ret < 0)
+			status = connector_status_unknown;
+	}
+
+	return status;
 }
 
 static const struct drm_edid *
@@ -664,12 +1283,47 @@ static int dw_hdmi_qp_bridge_scrambler_disable(struct drm_bridge *bridge)
 	return 0;
 }
 
+static enum drm_mode_status
+dw_hdmi_qp_bridge_frl_rate_valid(const struct drm_bridge *bridge,
+				 const struct drm_display_mode *mode,
+				 unsigned long long frl_data_rate,
+				 unsigned long long frl_max_link_rate)
+{
+	struct dw_hdmi_qp *hdmi = bridge->driver_private;
+	const struct dw_hdmi_qp_link_cfg *link_cfg;
+	unsigned long long frl_bw;
+
+	if (!hdmi->phy.ops->set_frl_rate) {
+		dev_dbg(hdmi->dev, "Unsupported FRL rate: %lld\n", frl_data_rate);
+		return MODE_CLOCK_HIGH;
+	}
+
+	link_cfg = hdmi->phy.ops->get_link_cfg(hdmi, hdmi->phy.data);
+
+	frl_bw = link_cfg->max_frl_rate_per_lane *
+		 link_cfg->max_frl_lanes * 1000000000ULL;
+	if (frl_bw < frl_data_rate) {
+		dev_dbg(hdmi->dev, "Unsupported FRL data rate: req=%llu max=%llu\n",
+			frl_data_rate, frl_bw);
+		return MODE_CLOCK_HIGH;
+	}
+
+	frl_bw = link_cfg->min_frl_rate_per_lane *
+		 link_cfg->min_frl_lanes * 1000000000ULL;
+	if (frl_bw > frl_max_link_rate) {
+		dev_dbg(hdmi->dev, "Unsupported FRL link rate: req=%llu min=%llu\n",
+			frl_max_link_rate, frl_bw);
+		return MODE_CLOCK_LOW;
+	}
+
+	return MODE_OK;
+}
+
 static int dw_hdmi_qp_bridge_clear_avi_infoframe(struct drm_bridge *bridge)
 {
 	struct dw_hdmi_qp *hdmi = bridge->driver_private;
 
-	dw_hdmi_qp_mod(hdmi, 0, PKTSCHED_AVI_TX_EN | PKTSCHED_GCP_TX_EN,
-		       PKTSCHED_PKT_EN);
+	dw_hdmi_qp_mod(hdmi, 0, PKTSCHED_AVI_TX_EN, PKTSCHED_PKT_EN);
 
 	return 0;
 }
@@ -750,8 +1404,8 @@ static int dw_hdmi_qp_bridge_write_avi_infoframe(struct drm_bridge *bridge,
 	dw_hdmi_qp_write_infoframe(hdmi, buffer, len, PKT_AVI_CONTENTS0);
 
 	dw_hdmi_qp_mod(hdmi, 0, PKTSCHED_AVI_FIELDRATE, PKTSCHED_PKT_CONFIG1);
-	dw_hdmi_qp_mod(hdmi, PKTSCHED_AVI_TX_EN | PKTSCHED_GCP_TX_EN,
-		       PKTSCHED_AVI_TX_EN | PKTSCHED_GCP_TX_EN, PKTSCHED_PKT_EN);
+	dw_hdmi_qp_mod(hdmi, PKTSCHED_AVI_TX_EN, PKTSCHED_AVI_TX_EN,
+		       PKTSCHED_PKT_EN);
 
 	return 0;
 }
@@ -1033,12 +1687,13 @@ static const struct drm_bridge_funcs dw_hdmi_qp_bridge_funcs = {
 	.atomic_reset = drm_atomic_helper_bridge_reset,
 	.atomic_enable = dw_hdmi_qp_bridge_atomic_enable,
 	.atomic_disable = dw_hdmi_qp_bridge_atomic_disable,
-	.detect = dw_hdmi_qp_bridge_detect,
+	.detect_ctx = dw_hdmi_qp_bridge_detect_ctx,
 	.hpd_enable = dw_hdmi_qp_bridge_hpd_enable,
 	.hpd_disable = dw_hdmi_qp_bridge_hpd_disable,
 	.edid_read = dw_hdmi_qp_bridge_edid_read,
 	.hdmi_scrambler_enable = dw_hdmi_qp_bridge_scrambler_enable,
 	.hdmi_scrambler_disable = dw_hdmi_qp_bridge_scrambler_disable,
+	.hdmi_frl_rate_valid = dw_hdmi_qp_bridge_frl_rate_valid,
 	.hdmi_clear_avi_infoframe = dw_hdmi_qp_bridge_clear_avi_infoframe,
 	.hdmi_write_avi_infoframe = dw_hdmi_qp_bridge_write_avi_infoframe,
 	.hdmi_clear_hdmi_infoframe = dw_hdmi_qp_bridge_clear_hdmi_infoframe,
@@ -1113,7 +1768,8 @@ struct dw_hdmi_qp *dw_hdmi_qp_bind(struct platform_device *pdev,
 	int ret;
 
 	if (!plat_data->phy_ops || !plat_data->phy_ops->init ||
-	    !plat_data->phy_ops->disable || !plat_data->phy_ops->read_hpd) {
+	    !plat_data->phy_ops->disable || !plat_data->phy_ops->read_hpd ||
+	    !plat_data->phy_ops->get_link_cfg) {
 		dev_err(dev, "Missing platform PHY ops\n");
 		return ERR_PTR(-ENODEV);
 	}
@@ -1123,6 +1779,8 @@ struct dw_hdmi_qp *dw_hdmi_qp_bind(struct platform_device *pdev,
 	if (IS_ERR(hdmi))
 		return ERR_CAST(hdmi);
 
+	INIT_WORK(&hdmi->flt_work, dw_hdmi_qp_flt_work);
+
 	hdmi->dev = dev;
 
 	regs = devm_platform_ioremap_resource(pdev, 0);
diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.h b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.h
index c07847e8d7dd0..dfb8cb6614ad7 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.h
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.h
@@ -23,6 +23,7 @@
 #define GLOBAL_SWRESET_REQUEST				0x40
 #define EARCRX_CMDC_SWINIT_P				BIT(27)
 #define AVP_DATAPATH_PACKET_AUDIO_SWINIT_P		BIT(10)
+#define AVP_DATAPATH_SWINIT_P				BIT(6)
 #define GLOBAL_SWDISABLE				0x44
 #define CEC_SWDISABLE					BIT(17)
 #define AVP_DATAPATH_PACKET_AUDIO_SWDISABLE		BIT(10)
@@ -216,6 +217,8 @@
 #define PKTSCHED_ACR_TX_EN				BIT(1)
 #define PKTSCHED_NULL_TX_EN				BIT(0)
 #define PKTSCHED_PKT_CONTROL0				0xaac
+#define PKTSCHED_GCP_CLEAR_AVMUTE			BIT(1)
+#define PKTSCHED_GCP_SET_AVMUTE				BIT(0)
 #define PKTSCHED_PKT_SEND				0xab0
 #define PKTSCHED_PKT_STATUS0				0xab4
 #define PKTSCHED_PKT_STATUS1				0xab8
diff --git a/drivers/gpu/drm/display/drm_bridge_connector.c b/drivers/gpu/drm/display/drm_bridge_connector.c
index a39a1902b1bcf..42b86cd867e4c 100644
--- a/drivers/gpu/drm/display/drm_bridge_connector.c
+++ b/drivers/gpu/drm/display/drm_bridge_connector.c
@@ -291,8 +291,11 @@ static int drm_bridge_connector_detect_ctx(struct drm_connector *connector,
 	int ret;
 
 	if (detect) {
-		status = detect->funcs->detect(detect, connector);
+		ret = drm_bridge_detect_ctx(detect, connector, ctx);
+		if (ret < 0)
+			return ret;
 
+		status = ret;
 		if (hdmi) {
 			ret = drm_atomic_helper_connector_hdmi_hotplug(connector, ctx, status);
 			if (ret == -EDEADLK)
@@ -448,6 +451,27 @@ drm_bridge_connector_tmds_char_rate_valid(const struct drm_connector *connector,
 		return MODE_OK;
 }
 
+static enum drm_mode_status
+drm_bridge_connector_frl_rate_valid(const struct drm_connector *connector,
+				    const struct drm_display_mode *mode,
+				    unsigned long long frl_data_rate,
+				    unsigned long long frl_max_link_rate)
+{
+	struct drm_bridge_connector *bridge_connector =
+		to_drm_bridge_connector(connector);
+	struct drm_bridge *bridge;
+
+	bridge = bridge_connector->bridge_hdmi;
+	if (!bridge)
+		return MODE_ERROR;
+
+	if (bridge->funcs->hdmi_frl_rate_valid)
+		return bridge->funcs->hdmi_frl_rate_valid(bridge, mode,
+							  frl_data_rate,
+							  frl_max_link_rate);
+	return MODE_CLOCK_HIGH;
+}
+
 static int drm_bridge_connector_clear_avi_infoframe(struct drm_connector *connector)
 {
 	struct drm_bridge_connector *bridge_connector =
@@ -627,6 +651,7 @@ static const struct drm_connector_hdmi_funcs drm_bridge_connector_hdmi_funcs = {
 	.supported_formats = BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444),
 	.max_bpc = 8,
 	.tmds_char_rate_valid = drm_bridge_connector_tmds_char_rate_valid,
+	.frl_rate_valid = drm_bridge_connector_frl_rate_valid,
 	.read_edid = drm_bridge_connector_read_edid,
 	.avi = {
 		.clear_infoframe = drm_bridge_connector_clear_avi_infoframe,
diff --git a/drivers/gpu/drm/display/drm_hdmi_state_helper.c b/drivers/gpu/drm/display/drm_hdmi_state_helper.c
index 27d85b1a50c7e..80cf37211bbf2 100644
--- a/drivers/gpu/drm/display/drm_hdmi_state_helper.c
+++ b/drivers/gpu/drm/display/drm_hdmi_state_helper.c
@@ -559,8 +559,30 @@ hdmi_clock_valid(const struct drm_connector *connector,
 	const struct drm_connector_hdmi_funcs *funcs = connector->hdmi.funcs;
 	const struct drm_display_info *info = &connector->display_info;
 
-	if (info->max_tmds_clock && clock > info->max_tmds_clock * 1000)
-		return MODE_CLOCK_HIGH;
+	if (info->max_tmds_clock && clock > info->max_tmds_clock * 1000) {
+		unsigned long long req_frl_bw, max_frl_bw;
+
+		/* Check if FRL is supported by the sink */
+		if (!info->hdmi.max_frl_rate_per_lane || !info->hdmi.max_lanes)
+			return MODE_CLOCK_HIGH;
+
+		/* Mandate further (i.e. driver) FRL checks */
+		if (!funcs || !funcs->frl_rate_valid)
+			return MODE_CLOCK_HIGH;
+
+		/*
+		 * Check required FRL bandwidth assuming 16b/18b encoding:
+		 * bw (bps) = clock (Hz) * 3 (channels) * 8 (bit) * 18 / 16
+		 */
+		req_frl_bw = clock * 27;
+		max_frl_bw = info->hdmi.max_frl_rate_per_lane *
+			     info->hdmi.max_lanes * 1000000000ULL;
+
+		if (req_frl_bw > max_frl_bw)
+			return MODE_CLOCK_HIGH;
+
+		return funcs->frl_rate_valid(connector, mode, req_frl_bw, max_frl_bw);
+	}
 
 	if (connector->hdmi.max_tmds_char_rate &&
 	    clock > connector->hdmi.max_tmds_char_rate)
diff --git a/drivers/gpu/drm/display/drm_scdc_helper.c b/drivers/gpu/drm/display/drm_scdc_helper.c
index 74388481fc556..7e137828c4fae 100644
--- a/drivers/gpu/drm/display/drm_scdc_helper.c
+++ b/drivers/gpu/drm/display/drm_scdc_helper.c
@@ -595,3 +595,188 @@ void drm_scdc_debugfs_init(struct drm_connector *connector, struct dentry *root)
 	debugfs_create_file("scdc_status", 0444, root, connector, &scdc_status_fops);
 }
 EXPORT_SYMBOL(drm_scdc_debugfs_init);
+
+static int drm_scdc_frl_config_to_rate(u8 config, u8 *rate_per_lane, u8 *lanes)
+{
+	switch (config) {
+	case SCDC_FRL_RATE_12GBPS_4LANE:
+		*rate_per_lane = 12;
+		*lanes = 4;
+		return 0;
+	case SCDC_FRL_RATE_10GBPS_4LANE:
+		*rate_per_lane = 10;
+		*lanes = 4;
+		return 0;
+	case SCDC_FRL_RATE_8GBPS_4LANE:
+		*rate_per_lane = 8;
+		*lanes = 4;
+		return 0;
+	case SCDC_FRL_RATE_6GBPS_4LANE:
+		*rate_per_lane = 6;
+		*lanes = 4;
+		return 0;
+	case SCDC_FRL_RATE_6GBPS_3LANE:
+		*rate_per_lane = 6;
+		*lanes = 3;
+		return 0;
+	case SCDC_FRL_RATE_3GBPS_3LANE:
+		*rate_per_lane = 3;
+		*lanes = 3;
+		return 0;
+	case SCDC_FRL_RATE_DISABLE:
+		*rate_per_lane = 0;
+		*lanes = 0;
+		return 0;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int drm_scdc_frl_rate_to_config(u8 rate_per_lane, u8 lanes)
+{
+	if (lanes != 0 && lanes != 3 && lanes != 4)
+		return -EINVAL;
+
+	switch (rate_per_lane * lanes) {
+	case 48:
+		return SCDC_FRL_RATE_12GBPS_4LANE;
+	case 40:
+		return SCDC_FRL_RATE_10GBPS_4LANE;
+	case 32:
+		return SCDC_FRL_RATE_8GBPS_4LANE;
+	case 24:
+		return SCDC_FRL_RATE_6GBPS_4LANE;
+	case 18:
+		return SCDC_FRL_RATE_6GBPS_3LANE;
+	case 9:
+		return SCDC_FRL_RATE_3GBPS_3LANE;
+	case 0:
+		return SCDC_FRL_RATE_DISABLE;
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * drm_scdc_set_frl - set FRL rate and FFE
+ * @connector: connector
+ * @rate_per_lane: FRL rate for a single lane (Gbps)
+ * @lanes: FRL lane count (3 or 4)
+ * @max_ffe_level: max TxFFE level for indicated FRL Rate (0..3)
+ *
+ * Writes over SCDC the FRL config register over SCDC channel, and sets
+ * FRL_Rate according to rate_per_lane x lanes, as well as FFE_levels
+ * according to max_ffe_level.
+ *
+ * Returns:
+ * True if write is successful, false otherwise.
+ */
+bool drm_scdc_set_frl(struct drm_connector *connector,
+		      u8 rate_per_lane, u8 lanes, u8 max_ffe_level)
+{
+	u8 config;
+	int ret;
+
+	ret = drm_scdc_frl_rate_to_config(rate_per_lane, lanes);
+	if (ret < 0 || max_ffe_level > 3) {
+		drm_dbg_kms(connector->dev,
+			    "[CONNECTOR:%d:%s] Invalid FRL config: rate=%ux%u ffe=%u\n",
+			    connector->base.id, connector->name, rate_per_lane,
+			    lanes, max_ffe_level);
+		return false;
+	}
+
+	config = FIELD_PREP(SCDC_FRL_RATE_MASK, ret) |
+		 FIELD_PREP(SCDC_FFE_LEVELS_MASK, max_ffe_level);
+
+	ret = drm_scdc_writeb(connector->ddc, SCDC_CONFIG_1, config);
+	if (ret < 0) {
+		drm_dbg_kms(connector->dev,
+			    "[CONNECTOR:%d:%s] Failed to set FRL: %d\n",
+			    connector->base.id, connector->name, ret);
+		return false;
+	}
+
+	return true;
+}
+EXPORT_SYMBOL(drm_scdc_set_frl);
+
+/**
+ * drm_scdc_calc_lower_frl - compute a reduced bandwidth FRL rate
+ * @in_rate_per_lane: input FRL rate for a single lane (Gbps)
+ * @in_lanes: input FRL lane count (3 or 4)
+ * @out_rate_per_lane: output FRL rate for a single lane (Gbps)
+ * @out_lanes: output FRL lane count (3 or 4)
+ *
+ * Determinates the FRL rate configuration with the highest bandwidth that is
+ * still lower than the bandwidth corresponding to the given input configuration.
+ * The resulting configuration is stored in out_rate_per_lane and out_lanes.
+ *
+ * Returns:
+ * True if computation was successful, false otherwise.
+ */
+bool drm_scdc_calc_lower_frl(u8 in_rate_per_lane, u8 in_lanes,
+			     u8 *out_rate_per_lane, u8 *out_lanes)
+{
+	int ret;
+
+	ret = drm_scdc_frl_rate_to_config(in_rate_per_lane, in_lanes);
+	if (ret < 0)
+		return false;
+
+	ret--;
+
+	if (ret <= SCDC_FRL_RATE_DISABLE)
+		return false;
+
+	ret = drm_scdc_frl_config_to_rate(ret, out_rate_per_lane, out_lanes);
+
+	return ret == 0;
+}
+EXPORT_SYMBOL(drm_scdc_calc_lower_frl);
+
+/**
+ * drm_scdc_get_frl_ltp_request - read LTP requested by the Sink for FRL lanes
+ * @connector: connector
+ * @ln0: output LTP request for FRL lane 0
+ * @ln1: output LTP request for FRL lane 1
+ * @ln2: output LTP request for FRL lane 2
+ * @ln3: output LTP request for FRL lane 3
+ *
+ * Reads over SCDC the Link Training Pattern (LTP) requested by the Sink for
+ * each of the four FRL lanes and stores the values in ln0..3.
+ *
+ * Returns:
+ * True on success, false otherwise.
+ */
+bool drm_scdc_get_frl_ltp_request(struct drm_connector *connector,
+				  u8 *ln0, u8 *ln1, u8 *ln2, u8 *ln3)
+{
+	u8 status;
+	int ret;
+
+	ret = drm_scdc_readb(connector->ddc, SCDC_STATUS_FLAGS_1, &status);
+	if (ret < 0) {
+		drm_dbg_kms(connector->dev,
+			    "[CONNECTOR:%d:%s] Failed to read LTP{0,1}: %d\n",
+			    connector->base.id, connector->name, ret);
+		return false;
+	}
+
+	*ln0 = FIELD_GET(SCDC_FRL_LN0_LTP_REQ_MASK, status);
+	*ln1 = FIELD_GET(SCDC_FRL_LN1_LTP_REQ_MASK, status);
+
+	ret = drm_scdc_readb(connector->ddc, SCDC_STATUS_FLAGS_2, &status);
+	if (ret < 0) {
+		drm_dbg_kms(connector->dev,
+			    "[CONNECTOR:%d:%s] Failed to read LTP{2,3}: %d\n",
+			    connector->base.id, connector->name, ret);
+		return false;
+	}
+
+	*ln2 = FIELD_GET(SCDC_FRL_LN2_LTP_REQ_MASK, status);
+	*ln3 = FIELD_GET(SCDC_FRL_LN3_LTP_REQ_MASK, status);
+
+	return true;
+}
+EXPORT_SYMBOL(drm_scdc_get_frl_ltp_request);
diff --git a/drivers/gpu/drm/drm_bridge.c b/drivers/gpu/drm/drm_bridge.c
index 3bfa0e911b09c..a46302f446d7e 100644
--- a/drivers/gpu/drm/drm_bridge.c
+++ b/drivers/gpu/drm/drm_bridge.c
@@ -1472,9 +1472,9 @@ EXPORT_SYMBOL(drm_atomic_bridge_chain_check);
  * @connector: attached connector
  *
  * If the bridge supports output detection, as reported by the
- * DRM_BRIDGE_OP_DETECT bridge ops flag, call &drm_bridge_funcs.detect for the
- * bridge and return the connection status. Otherwise return
- * connector_status_unknown.
+ * DRM_BRIDGE_OP_DETECT bridge ops flag, call &drm_bridge_funcs.detect_ctx
+ * or &drm_bridge_funcs.detect for the bridge and return the connection status.
+ * Otherwise return connector_status_unknown.
  *
  * RETURNS:
  * The detection status on success, or connector_status_unknown if the bridge
@@ -1483,12 +1483,68 @@ EXPORT_SYMBOL(drm_atomic_bridge_chain_check);
 enum drm_connector_status
 drm_bridge_detect(struct drm_bridge *bridge, struct drm_connector *connector)
 {
+	return drm_bridge_detect_ctx(bridge, connector, NULL);
+}
+EXPORT_SYMBOL_GPL(drm_bridge_detect);
+
+/**
+ * drm_bridge_detect_ctx - check if anything is attached to the bridge output
+ * @bridge: bridge control structure
+ * @connector: attached connector
+ * @ctx: acquire_ctx, or NULL to let this function handle locking
+ *
+ * If the bridge supports output detection, as reported by the
+ * DRM_BRIDGE_OP_DETECT bridge ops flag, call &drm_bridge_funcs.detect_ctx
+ * or &drm_bridge_funcs.detect for the bridge and return the connection status.
+ * Otherwise return connector_status_unknown.
+ *
+ * RETURNS:
+ * The detection status on success, or connector_status_unknown if the bridge
+ * doesn't support output detection.
+ * If @ctx is set, it might also return -EDEADLK.
+ */
+int drm_bridge_detect_ctx(struct drm_bridge *bridge,
+			  struct drm_connector *connector,
+			  struct drm_modeset_acquire_ctx *ctx)
+{
+	struct drm_modeset_acquire_ctx br_ctx;
+	int ret;
+
 	if (!(bridge->ops & DRM_BRIDGE_OP_DETECT))
 		return connector_status_unknown;
 
-	return bridge->funcs->detect(bridge, connector);
+	if (!bridge->funcs->detect_ctx)
+		return bridge->funcs->detect(bridge, connector);
+
+	if (ctx) {
+		ret = bridge->funcs->detect_ctx(bridge, connector, ctx);
+		if (ret == -EDEADLK)
+			return ret;
+
+		goto out;
+	}
+
+	drm_modeset_acquire_init(&br_ctx, 0);
+retry:
+	ret = drm_modeset_lock(&connector->dev->mode_config.connection_mutex,
+			       &br_ctx);
+	if (!ret)
+		ret = bridge->funcs->detect_ctx(bridge, connector, &br_ctx);
+
+	if (ret == -EDEADLK) {
+		drm_modeset_backoff(&br_ctx);
+		goto retry;
+	}
+
+	drm_modeset_drop_locks(&br_ctx);
+	drm_modeset_acquire_fini(&br_ctx);
+out:
+	if (WARN_ON(ret < 0))
+		ret = connector_status_unknown;
+
+	return ret;
 }
-EXPORT_SYMBOL_GPL(drm_bridge_detect);
+EXPORT_SYMBOL_GPL(drm_bridge_detect_ctx);
 
 /**
  * drm_bridge_get_modes - fill all modes currently valid for the sink into the
diff --git a/drivers/gpu/drm/rockchip/Kconfig b/drivers/gpu/drm/rockchip/Kconfig
index e7f49fe845eae..12188ad1fb4aa 100644
--- a/drivers/gpu/drm/rockchip/Kconfig
+++ b/drivers/gpu/drm/rockchip/Kconfig
@@ -20,7 +20,7 @@ config DRM_ROCKCHIP
 	select DRM_INNO_HDMI if ROCKCHIP_INNO_HDMI
 	select GENERIC_PHY if ROCKCHIP_DW_MIPI_DSI
 	select GENERIC_PHY_MIPI_DPHY if ROCKCHIP_DW_MIPI_DSI
-	select SND_SOC_HDMI_CODEC if ROCKCHIP_CDN_DP && SND_SOC
+	select SND_SOC_HDMI_CODEC if SND_SOC && (ROCKCHIP_CDN_DP || ROCKCHIP_DW_HDMI_QP || ROCKCHIP_DW_DP)
 	help
 	  Choose this option if you have a Rockchip soc chipset.
 	  This driver provides kernel mode setting and buffer
diff --git a/drivers/gpu/drm/rockchip/dw_hdmi_qp-rockchip.c b/drivers/gpu/drm/rockchip/dw_hdmi_qp-rockchip.c
index 8b0a71a3709f4..431658f89d812 100644
--- a/drivers/gpu/drm/rockchip/dw_hdmi_qp-rockchip.c
+++ b/drivers/gpu/drm/rockchip/dw_hdmi_qp-rockchip.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 /*
  * Copyright (c) 2021-2022 Rockchip Electronics Co., Ltd.
- * Copyright (c) 2024 Collabora Ltd.
+ * Copyright (c) 2024-2026 Collabora Ltd.
  *
  * Author: Algea Cao <algea.cao@rock-chips.com>
  * Author: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
@@ -59,6 +59,7 @@
 #define RK3588_HDMI0_HPD_INT_CLR	BIT(12)
 #define RK3588_HDMI1_HPD_INT_MSK	BIT(15)
 #define RK3588_HDMI1_HPD_INT_CLR	BIT(14)
+
 #define RK3588_GRF_SOC_CON7		0x031c
 #define RK3588_HPD_HDMI0_IO_EN_MASK	BIT(12)
 #define RK3588_HPD_HDMI1_IO_EN_MASK	BIT(13)
@@ -67,7 +68,11 @@
 #define RK3588_HDMI0_OHPD_INT		BIT(17)
 #define RK3588_HDMI1_LEVEL_INT		BIT(24)
 #define RK3588_HDMI1_OHPD_INT		BIT(25)
+
 #define RK3588_GRF_VO1_CON3		0x000c
+#define RK3588_GRF_VO1_CON4		0x0010
+#define RK3588_HDMI21_MASK		BIT(0)
+
 #define RK3588_GRF_VO1_CON6		0x0018
 #define RK3588_COLOR_DEPTH_MASK		GENMASK(7, 4)
 #define RK3588_8BPC			0x0
@@ -77,12 +82,19 @@
 #define RK3588_SDAIN_MASK		BIT(10)
 #define RK3588_MODE_MASK		BIT(11)
 #define RK3588_I2S_SEL_MASK		BIT(13)
+
+#define RK3588_GRF_VO1_CON7		0x001c
 #define RK3588_GRF_VO1_CON9		0x0024
 #define RK3588_HDMI0_GRANT_SEL		BIT(10)
 #define RK3588_HDMI1_GRANT_SEL		BIT(12)
 
 #define HOTPLUG_DEBOUNCE_MS		150
 #define MAX_HDMI_PORT_NUM		2
+#define HDMI20_MAX_TMDS_RATE		600000000
+#define HDMI21_MAX_FRL_LANE_RATE	12
+#define HDMI21_MAX_FRL_LANE_NUM		4
+#define HDMI21_MIN_FRL_LANE_RATE	3
+#define HDMI21_MIN_FRL_LANE_NUM		3
 
 #define RK_COLOR_FMT_RGB		0x0
 #define RK_COLOR_FMT_YUV422		0x1
@@ -96,12 +108,12 @@ struct rockchip_hdmi_qp {
 	struct rockchip_encoder encoder;
 	struct dw_hdmi_qp *hdmi;
 	struct phy *phy;
+	struct dw_hdmi_qp_link_cfg link_cfg;
 	struct gpio_desc *frl_enable_gpio;
 	struct delayed_work hpd_work;
 	int hpd_irq;
 	int port_id;
 	const struct rockchip_hdmi_qp_ctrl_ops *ctrl_ops;
-	unsigned long long tmds_char_rate;
 };
 
 struct rockchip_hdmi_qp_ctrl_ops {
@@ -148,16 +160,24 @@ static int __pure dw_hdmi_qp_rockchip_bus_fmt_to_reg(u32 bus_fmt)
 static void dw_hdmi_qp_rockchip_encoder_enable(struct drm_encoder *encoder)
 {
 	struct rockchip_hdmi_qp *hdmi = to_rockchip_hdmi_qp(encoder);
+	const struct dw_hdmi_qp_link_cfg *lcfg = &hdmi->link_cfg;
 	struct drm_crtc *crtc = encoder->crtc;
+	struct rockchip_crtc_state *rks;
 
-	/* Unconditionally switch to TMDS as FRL is not yet supported */
-	gpiod_set_value_cansleep(hdmi->frl_enable_gpio, 0);
+	gpiod_set_value_cansleep(hdmi->frl_enable_gpio, lcfg->frl_enabled);
 
 	if (!crtc || !crtc->state)
 		return;
 
+	rks = to_rockchip_crtc_state(crtc->state);
+
 	if (hdmi->ctrl_ops->enc_init)
-		hdmi->ctrl_ops->enc_init(hdmi, to_rockchip_crtc_state(crtc->state));
+		hdmi->ctrl_ops->enc_init(hdmi, rks);
+
+	dev_dbg(hdmi->dev, "%s port=%d tmds=%llu frl=%ux%u bpc=%u/%u\n",
+		__func__, hdmi->port_id, lcfg->tmds_char_rate,
+		lcfg->frl_rate_per_lane, lcfg->frl_lanes,
+		rks->output_bpc, lcfg->bpc);
 }
 
 /**
@@ -196,11 +216,14 @@ dw_hdmi_qp_rockchip_encoder_atomic_check(struct drm_encoder *encoder,
 					 struct drm_connector_state *conn_state)
 {
 	const struct drm_display_mode *adj_mode = &crtc_state->adjusted_mode;
+	const struct drm_display_info *info = &conn_state->connector->display_info;
 	struct rockchip_crtc_state *s = to_rockchip_crtc_state(crtc_state);
 	struct rockchip_hdmi_qp *hdmi = to_rockchip_hdmi_qp(encoder);
+	struct dw_hdmi_qp_link_cfg *lcfg = &hdmi->link_cfg;
 	union phy_configure_opts phy_cfg = {};
 	unsigned int overscan;
 	int ingest_fmt;
+	enum phy_hdmi_mode mode;
 	int ret;
 
 	ingest_fmt = dw_hdmi_qp_rockchip_get_vop_format(encoder, conn_state);
@@ -213,7 +236,7 @@ dw_hdmi_qp_rockchip_encoder_atomic_check(struct drm_encoder *encoder,
 	s->tv_margins.top    = adj_mode->vdisplay * overscan / 200;
 	s->tv_margins.bottom = s->tv_margins.top;
 
-	if (hdmi->tmds_char_rate == conn_state->hdmi.tmds_char_rate &&
+	if (lcfg->tmds_char_rate == conn_state->hdmi.tmds_char_rate &&
 	    s->output_bpc == conn_state->hdmi.output_bpc &&
 	    s->bus_format == ingest_fmt)
 		return 0;
@@ -237,7 +260,6 @@ dw_hdmi_qp_rockchip_encoder_atomic_check(struct drm_encoder *encoder,
 		return -EINVAL;
 	}
 
-	phy_cfg.hdmi.tmds_char_rate = conn_state->hdmi.tmds_char_rate;
 	/*
 	 * YUV422 always transmits two 12-bit components per clock cycle,
 	 * regardless of the color depth, which from a rate perspective is
@@ -247,17 +269,58 @@ dw_hdmi_qp_rockchip_encoder_atomic_check(struct drm_encoder *encoder,
 	phy_cfg.hdmi.bpc = (s->output_mode == ROCKCHIP_OUT_MODE_YUV422 ?
 				8 : conn_state->hdmi.output_bpc);
 
-	ret = phy_configure(hdmi->phy, &phy_cfg);
-	if (!ret) {
-		hdmi->tmds_char_rate = conn_state->hdmi.tmds_char_rate;
-		s->output_type = DRM_MODE_CONNECTOR_HDMIA;
-		s->output_bpc = conn_state->hdmi.output_bpc;
-		s->bus_format = ingest_fmt;
+	if (conn_state->hdmi.tmds_char_rate > HDMI20_MAX_TMDS_RATE ||
+	    (info->max_tmds_clock &&
+	     conn_state->hdmi.tmds_char_rate > info->max_tmds_clock * 1000)) {
+		mode = PHY_HDMI_MODE_FRL;
+
+		if (info->hdmi.max_frl_rate_per_lane > lcfg->max_frl_rate_per_lane)
+			phy_cfg.hdmi.frl.rate_per_lane = lcfg->max_frl_rate_per_lane;
+		else
+			phy_cfg.hdmi.frl.rate_per_lane = info->hdmi.max_frl_rate_per_lane;
+
+		if (info->hdmi.max_lanes > lcfg->max_frl_lanes)
+			phy_cfg.hdmi.frl.lanes = lcfg->max_frl_lanes;
+		else
+			phy_cfg.hdmi.frl.lanes = info->hdmi.max_lanes;
 	} else {
+		mode = PHY_HDMI_MODE_TMDS;
+
+		phy_cfg.hdmi.tmds_char_rate = conn_state->hdmi.tmds_char_rate;
+	}
+
+	ret = phy_set_mode_ext(hdmi->phy, PHY_MODE_HDMI, mode);
+	if (ret) {
+		dev_err(hdmi->dev, "Failed to switch phy mode: %d\n", ret);
+		return ret;
+	}
+
+	ret = phy_configure(hdmi->phy, &phy_cfg);
+	if (ret) {
 		dev_err(hdmi->dev, "Failed to configure phy: %d\n", ret);
+		return ret;
 	}
 
-	return ret;
+	lcfg->tmds_char_rate = conn_state->hdmi.tmds_char_rate;
+
+	if (mode == PHY_HDMI_MODE_FRL) {
+		lcfg->frl_enabled = true;
+		lcfg->frl_rate_per_lane = phy_cfg.hdmi.frl.rate_per_lane;
+		lcfg->frl_lanes = phy_cfg.hdmi.frl.lanes;
+	} else {
+		lcfg->frl_enabled = false;
+		lcfg->frl_rate_per_lane = 0;
+		lcfg->frl_lanes = 0;
+	}
+
+	lcfg->bpc = phy_cfg.hdmi.bpc;
+
+	s->output_type = DRM_MODE_CONNECTOR_HDMIA;
+	s->output_bpc = conn_state->hdmi.output_bpc;
+	s->bus_format = ingest_fmt;
+	s->frl_enabled = lcfg->frl_enabled;
+
+	return 0;
 }
 
 static const struct
@@ -335,12 +398,68 @@ static void dw_hdmi_qp_rk3588_disable_hpd(struct dw_hdmi_qp *dw_hdmi, void *data
 	dw_hdmi_qp_rk3588_mask_hpd(hdmi);
 }
 
+static const struct dw_hdmi_qp_link_cfg *
+dw_hdmi_qp_rk3588_get_link_cfg(struct dw_hdmi_qp *dw_hdmi, void *data)
+{
+	struct rockchip_hdmi_qp *hdmi = (struct rockchip_hdmi_qp *)data;
+
+	return &hdmi->link_cfg;
+}
+
+static int dw_hdmi_qp_rk3588_set_frl_rate(struct dw_hdmi_qp *dw_hdmi, void *data,
+					  u8 rate_per_lane, u8 lanes)
+{
+	struct rockchip_hdmi_qp *hdmi = (struct rockchip_hdmi_qp *)data;
+	union phy_configure_opts phy_cfg = {};
+	int ret;
+
+	if (!hdmi->link_cfg.frl_enabled || !rate_per_lane || !lanes)
+		return -EINVAL;
+
+	phy_cfg.hdmi.frl.rate_per_lane = rate_per_lane;
+	phy_cfg.hdmi.frl.lanes = lanes;
+	phy_cfg.hdmi.bpc = hdmi->link_cfg.bpc;
+
+	ret = phy_configure(hdmi->phy, &phy_cfg);
+	if (ret) {
+		dev_err(hdmi->dev, "Failed to set PHY FRL rate: %d\n", ret);
+	} else {
+		hdmi->link_cfg.frl_rate_per_lane = rate_per_lane;
+		hdmi->link_cfg.frl_lanes = lanes;
+	}
+
+	return ret;
+}
+
+static int dw_hdmi_qp_rk3588_set_ffe_level(struct dw_hdmi_qp *dw_hdmi, void *data,
+					   u8 ffe_level)
+{
+	struct rockchip_hdmi_qp *hdmi = (struct rockchip_hdmi_qp *)data;
+	union phy_configure_opts phy_cfg = {};
+	int ret;
+
+	if (!hdmi->link_cfg.frl_enabled)
+		return -EINVAL;
+
+	phy_cfg.hdmi.frl.ffe_level = ffe_level;
+	phy_cfg.hdmi.frl.set_ffe_level = true;
+
+	ret = phy_configure(hdmi->phy, &phy_cfg);
+	if (ret)
+		dev_err(hdmi->dev, "Failed to set PHY FFE level: %d\n", ret);
+
+	return ret;
+}
+
 static const struct dw_hdmi_qp_phy_ops rk3588_hdmi_phy_ops = {
 	.init		= dw_hdmi_qp_rk3588_phy_init,
 	.disable	= dw_hdmi_qp_rk3588_phy_disable,
 	.read_hpd	= dw_hdmi_qp_rk3588_read_hpd,
 	.enable_hpd	= dw_hdmi_qp_rk3588_enable_hpd,
 	.disable_hpd	= dw_hdmi_qp_rk3588_disable_hpd,
+	.get_link_cfg	= dw_hdmi_qp_rk3588_get_link_cfg,
+	.set_frl_rate	= dw_hdmi_qp_rk3588_set_frl_rate,
+	.set_ffe_level	= dw_hdmi_qp_rk3588_set_ffe_level,
 };
 
 static enum drm_connector_status
@@ -390,6 +509,9 @@ static const struct dw_hdmi_qp_phy_ops rk3576_hdmi_phy_ops = {
 	.read_hpd	= dw_hdmi_qp_rk3576_read_hpd,
 	.enable_hpd	= dw_hdmi_qp_rk3576_enable_hpd,
 	.disable_hpd	= dw_hdmi_qp_rk3576_disable_hpd,
+	.get_link_cfg	= dw_hdmi_qp_rk3588_get_link_cfg,
+	.set_frl_rate	= dw_hdmi_qp_rk3588_set_frl_rate,
+	.set_ffe_level	= dw_hdmi_qp_rk3588_set_ffe_level,
 };
 
 static void dw_hdmi_qp_rk3588_hpd_work(struct work_struct *work)
@@ -503,6 +625,9 @@ static void dw_hdmi_qp_rk3576_enc_init(struct rockchip_hdmi_qp *hdmi,
 	int color = dw_hdmi_qp_rockchip_bus_fmt_to_reg(state->bus_format);
 	u32 val;
 
+	val = FIELD_PREP_WM16(RK3576_HDMI_FRL_MOD, hdmi->link_cfg.frl_enabled);
+	regmap_write(hdmi->vo_regmap, RK3576_VO0_GRF_SOC_CON1, val);
+
 	if (state->output_bpc == 10)
 		val = FIELD_PREP_WM16(RK3576_COLOR_DEPTH_MASK, RK3576_10BPC);
 	else
@@ -520,6 +645,11 @@ static void dw_hdmi_qp_rk3588_enc_init(struct rockchip_hdmi_qp *hdmi,
 	int color = dw_hdmi_qp_rockchip_bus_fmt_to_reg(state->bus_format);
 	u32 val;
 
+	val = FIELD_PREP_WM16(RK3588_HDMI21_MASK, hdmi->link_cfg.frl_enabled);
+	regmap_write(hdmi->vo_regmap,
+		     hdmi->port_id ? RK3588_GRF_VO1_CON7 : RK3588_GRF_VO1_CON4,
+		     val);
+
 	if (state->output_bpc == 10)
 		val = FIELD_PREP_WM16(RK3588_COLOR_DEPTH_MASK, RK3588_10BPC);
 	else
@@ -547,11 +677,24 @@ static const struct rockchip_hdmi_qp_ctrl_ops rk3588_hdmi_ctrl_ops = {
 	.hardirq_callback	= dw_hdmi_qp_rk3588_hardirq,
 };
 
+enum rockchip_hdmi_qp_ffe_cfg {
+	FFE_LEVEL_AUTO = 0,
+	FFE_LEVEL_0,
+	FFE_LEVEL_1,
+	FFE_LEVEL_2,
+	FFE_LEVEL_3,
+};
+
 struct rockchip_hdmi_qp_cfg {
 	unsigned int num_ports;
 	unsigned int port_ids[MAX_HDMI_PORT_NUM];
 	const struct rockchip_hdmi_qp_ctrl_ops *ctrl_ops;
 	const struct dw_hdmi_qp_phy_ops *phy_ops;
+	u8 max_frl_rate_per_lane;
+	u8 max_frl_lanes;
+	u8 min_frl_rate_per_lane;
+	u8 min_frl_lanes;
+	enum rockchip_hdmi_qp_ffe_cfg max_ffe;
 };
 
 static const struct rockchip_hdmi_qp_cfg rk3576_hdmi_cfg = {
@@ -571,6 +714,9 @@ static const struct rockchip_hdmi_qp_cfg rk3588_hdmi_cfg = {
 	},
 	.ctrl_ops = &rk3588_hdmi_ctrl_ops,
 	.phy_ops = &rk3588_hdmi_phy_ops,
+	/* FIXME: Intermittent screen flicker if rate exceeds 40 Gbps */
+	.max_frl_rate_per_lane = 10,
+	.max_frl_lanes = 4,
 };
 
 static const struct of_device_id dw_hdmi_qp_rockchip_dt_ids[] = {
@@ -585,12 +731,21 @@ static const struct of_device_id dw_hdmi_qp_rockchip_dt_ids[] = {
 };
 MODULE_DEVICE_TABLE(of, dw_hdmi_qp_rockchip_dt_ids);
 
+static u8 dw_hdmi_qp_rockchip_ffe_cfg_to_level(enum rockchip_hdmi_qp_ffe_cfg ffe)
+{
+	if (ffe == FFE_LEVEL_AUTO)
+		ffe = FFE_LEVEL_3;
+
+	return ffe - 1;
+}
+
 static int dw_hdmi_qp_rockchip_bind(struct device *dev, struct device *master,
 				    void *data)
 {
 	struct platform_device *pdev = to_platform_device(dev);
 	struct dw_hdmi_qp_plat_data plat_data = {};
 	const struct rockchip_hdmi_qp_cfg *cfg;
+	struct dw_hdmi_qp_link_cfg *lcfg;
 	struct drm_device *drm = data;
 	struct drm_connector *connector;
 	struct drm_encoder *encoder;
@@ -691,6 +846,26 @@ static int dw_hdmi_qp_rockchip_bind(struct device *dev, struct device *master,
 	if (IS_ERR(hdmi->phy))
 		return dev_err_probe(dev, PTR_ERR(hdmi->phy), "Failed to get phy\n");
 
+	lcfg = &hdmi->link_cfg;
+	lcfg->max_frl_rate_per_lane = cfg->max_frl_rate_per_lane ?: HDMI21_MAX_FRL_LANE_RATE;
+	lcfg->max_frl_lanes = cfg->max_frl_lanes ?: HDMI21_MAX_FRL_LANE_NUM;
+	lcfg->min_frl_rate_per_lane = cfg->min_frl_rate_per_lane ?: HDMI21_MIN_FRL_LANE_RATE;
+	lcfg->min_frl_lanes = cfg->min_frl_lanes ?: HDMI21_MIN_FRL_LANE_NUM;
+
+	if (lcfg->max_frl_rate_per_lane < lcfg->min_frl_rate_per_lane ||
+	    lcfg->max_frl_lanes < lcfg->min_frl_lanes ||
+	    lcfg->max_frl_rate_per_lane > HDMI21_MAX_FRL_LANE_RATE ||
+	    lcfg->max_frl_rate_per_lane < HDMI21_MIN_FRL_LANE_RATE ||
+	    lcfg->max_frl_lanes > HDMI21_MAX_FRL_LANE_NUM ||
+	    lcfg->max_frl_lanes < HDMI21_MIN_FRL_LANE_NUM ||
+	    lcfg->min_frl_rate_per_lane > HDMI21_MAX_FRL_LANE_RATE ||
+	    lcfg->min_frl_rate_per_lane < HDMI21_MIN_FRL_LANE_RATE ||
+	    lcfg->min_frl_lanes > HDMI21_MAX_FRL_LANE_NUM ||
+	    lcfg->min_frl_lanes < HDMI21_MIN_FRL_LANE_NUM)
+		return dev_err_probe(hdmi->dev, -EINVAL, "Invalid FRL config\n");
+
+	lcfg->max_ffe_level = dw_hdmi_qp_rockchip_ffe_cfg_to_level(cfg->max_ffe);
+
 	cfg->ctrl_ops->io_init(hdmi);
 
 	INIT_DELAYED_WORK(&hdmi->hpd_work, dw_hdmi_qp_rk3588_hpd_work);
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_drv.h b/drivers/gpu/drm/rockchip/rockchip_drm_drv.h
index d57620c66f16d..923345777a2fd 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_drv.h
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_drv.h
@@ -59,6 +59,7 @@ struct rockchip_crtc_state {
 	u32 bus_flags;
 	int color_space;
 	struct drm_connector_tv_margins tv_margins;
+	bool frl_enabled;
 };
 #define to_rockchip_crtc_state(s) \
 		container_of(s, struct rockchip_crtc_state, base)
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index e4cb229f84a33..f5d2f47c5dbf0 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -104,6 +104,8 @@ enum vop2_afbc_format {
 };
 
 #define VOP2_MAX_DCLK_RATE		600000000ULL
+#define VOP2_ACLK_RATE_DEFAULT		500000000UL
+#define VOP2_ACLK_RATE_FRL		750000000UL
 
 /*
  * bus-format types.
@@ -1092,6 +1094,17 @@ static void vop2_crtc_atomic_disable(struct drm_crtc *crtc,
 
 	vp->enabled = false;
 
+	if (vop2->version == VOP_VERSION_RK3588) {
+		struct rockchip_crtc_state *vcstate = to_rockchip_crtc_state(old_crtc_state);
+
+		if (vcstate->frl_enabled) {
+			vop2->frl_vp_count--;
+
+			if (!vop2->frl_vp_count)
+				clk_set_rate(vop2->aclk, VOP2_ACLK_RATE_DEFAULT);
+		}
+	}
+
 	vop2->enable_count--;
 
 	if (!vop2->enable_count)
@@ -1858,6 +1871,13 @@ static void vop2_crtc_atomic_enable(struct drm_crtc *crtc,
 
 	vop2->enable_count++;
 
+	if (vop2->version == VOP_VERSION_RK3588 && vcstate->frl_enabled) {
+		if (!vop2->frl_vp_count)
+			clk_set_rate(vop2->aclk, VOP2_ACLK_RATE_FRL);
+
+		vop2->frl_vp_count++;
+	}
+
 	vcstate->yuv_overlay = is_yuv_output(vcstate->bus_format);
 
 	vop2_crtc_enable_irq(vp, VP_INT_POST_BUF_EMPTY);
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h
index 52537278b21f3..f4fd0ea66bc95 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h
@@ -333,6 +333,7 @@ struct vop2 {
 	 * we need a ref counter here.
 	 */
 	unsigned int enable_count;
+	unsigned int frl_vp_count;
 	struct clk *hclk;
 	struct clk *aclk;
 	struct clk *pclk;
diff --git a/drivers/phy/rockchip/phy-rockchip-samsung-hdptx.c b/drivers/phy/rockchip/phy-rockchip-samsung-hdptx.c
index aa0e362603810..f03adeb8d0d66 100644
--- a/drivers/phy/rockchip/phy-rockchip-samsung-hdptx.c
+++ b/drivers/phy/rockchip/phy-rockchip-samsung-hdptx.c
@@ -333,6 +333,7 @@
 #define FRL_3G3L_RATE			900000000
 #define FRL_6G3L_RATE			1800000000
 #define FRL_8G4L_RATE			3200000000
+#define FRL_FFE_MAX_LEVEL		3
 
 enum dp_link_rate {
 	DP_BW_RBR,
@@ -466,6 +467,16 @@ static const struct ropll_config rk_hdptx_tmds_ropll_cfg[] = {
 	{  25175000ULL,   84,    84,    1,    1,   15,  1,  168,    1,  16,  4,   1,    1, },
 };
 
+static const struct ffe_config {
+	u8 pre_shoot;
+	u8 de_emphasis;
+} rk_hdptx_frl_ffe_cfg[FRL_FFE_MAX_LEVEL + 1] = {
+	{ 0x3, 0x4 },
+	{ 0x3, 0x6 },
+	{ 0x3, 0x8 },
+	{ 0x3, 0x9 },
+};
+
 static const struct reg_sequence rk_hdptx_common_cmn_init_seq[] = {
 	REG_SEQ0(CMN_REG(0009), 0x0c),
 	REG_SEQ0(CMN_REG(000a), 0x83),
@@ -1321,6 +1332,45 @@ static int rk_hdptx_tmds_ropll_mode_config(struct rk_hdptx_phy *hdptx)
 	return rk_hdptx_post_enable_lane(hdptx);
 }
 
+static int rk_hdptx_frl_ffe_config(struct rk_hdptx_phy *hdptx, u8 ffe_level)
+{
+	u8 val;
+
+	if (ffe_level > FRL_FFE_MAX_LEVEL)
+		return -EINVAL;
+
+	val = rk_hdptx_frl_ffe_cfg[ffe_level].pre_shoot;
+
+	regmap_update_bits(hdptx->regmap, LANE_REG(0305),
+			   LN_TX_DRV_PRE_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_PRE_LVL_CTRL_MASK, val));
+	regmap_update_bits(hdptx->regmap, LANE_REG(0405),
+			   LN_TX_DRV_PRE_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_PRE_LVL_CTRL_MASK, val));
+	regmap_update_bits(hdptx->regmap, LANE_REG(0505),
+			   LN_TX_DRV_PRE_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_PRE_LVL_CTRL_MASK, val));
+	regmap_update_bits(hdptx->regmap, LANE_REG(0605),
+			   LN_TX_DRV_PRE_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_PRE_LVL_CTRL_MASK, val));
+
+	val = rk_hdptx_frl_ffe_cfg[ffe_level].de_emphasis;
+
+	regmap_update_bits(hdptx->regmap, LANE_REG(0304),
+			   LN_TX_DRV_POST_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_POST_LVL_CTRL_MASK, val));
+	regmap_update_bits(hdptx->regmap, LANE_REG(0404),
+			   LN_TX_DRV_POST_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_POST_LVL_CTRL_MASK, val));
+	regmap_update_bits(hdptx->regmap, LANE_REG(0504),
+			   LN_TX_DRV_POST_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_POST_LVL_CTRL_MASK, val));
+	regmap_update_bits(hdptx->regmap, LANE_REG(0604),
+			   LN_TX_DRV_POST_LVL_CTRL_MASK,
+			   FIELD_PREP(LN_TX_DRV_POST_LVL_CTRL_MASK, val));
+	return 0;
+}
+
 static void rk_hdptx_dp_reset(struct rk_hdptx_phy *hdptx)
 {
 	reset_control_assert(hdptx->rsts[RST_LANE].rstc);
@@ -1735,6 +1785,13 @@ static int rk_hdptx_phy_verify_hdmi_config(struct rk_hdptx_phy *hdptx,
 		unsigned long long frl_rate = 100000000ULL * hdmi_in->frl.lanes *
 					      hdmi_in->frl.rate_per_lane;
 
+		if (hdmi_in->frl.set_ffe_level) {
+			if (hdmi_in->frl.ffe_level > FRL_FFE_MAX_LEVEL)
+				return -EINVAL;
+
+			return 0;
+		}
+
 		switch (hdmi_in->frl.rate_per_lane) {
 		case 3:
 		case 6:
@@ -2081,11 +2138,18 @@ static int rk_hdptx_phy_configure(struct phy *phy, union phy_configure_opts *opt
 		if (ret) {
 			dev_err(hdptx->dev, "invalid hdmi params for phy configure\n");
 		} else {
-			hdptx->pll_config_dirty = true;
-
-			dev_dbg(hdptx->dev, "%s %s rate=%llu bpc=%u\n", __func__,
-				hdptx->hdmi_cfg.mode ? "FRL" : "TMDS",
-				hdptx->hdmi_cfg.rate, hdptx->hdmi_cfg.bpc);
+			if (hdptx->hdmi_cfg.mode == PHY_HDMI_MODE_FRL &&
+			    opts->hdmi.frl.set_ffe_level) {
+				dev_dbg(hdptx->dev, "%s ffe_level=%u\n", __func__,
+					opts->hdmi.frl.ffe_level);
+				ret = rk_hdptx_frl_ffe_config(hdptx, opts->hdmi.frl.ffe_level);
+			} else {
+				hdptx->pll_config_dirty = true;
+
+				dev_dbg(hdptx->dev, "%s %s rate=%llu bpc=%u\n", __func__,
+					hdptx->hdmi_cfg.mode ? "FRL" : "TMDS",
+					hdptx->hdmi_cfg.rate, hdptx->hdmi_cfg.bpc);
+			}
 		}
 
 		return ret;
@@ -2337,14 +2401,16 @@ static int rk_hdptx_phy_clk_determine_rate(struct clk_hw *hw,
 	 * to ensure rk_hdptx_phy_clk_set_rate() will be always invoked.
 	 * Otherwise, restrict the rate according to the PHY link setup.
 	 */
-	if (hdptx->pll_config_dirty)
-		req->rate = 0;
-	else if (hdptx->hdmi_cfg.mode == PHY_HDMI_MODE_FRL)
+
+	if (hdptx->hdmi_cfg.mode == PHY_HDMI_MODE_FRL)
 		req->rate = hdptx->hdmi_cfg.rate;
 	else
 		req->rate = DIV_ROUND_CLOSEST_ULL(hdptx->hdmi_cfg.rate * 8,
 						  hdptx->hdmi_cfg.bpc);
 
+	if (hdptx->pll_config_dirty && req->rate == clk_hw_get_rate(hw))
+		req->rate = 0;
+
 	return 0;
 }
 
diff --git a/include/drm/bridge/dw_hdmi_qp.h b/include/drm/bridge/dw_hdmi_qp.h
index ad33f9cafdebd..87d473a0051a9 100644
--- a/include/drm/bridge/dw_hdmi_qp.h
+++ b/include/drm/bridge/dw_hdmi_qp.h
@@ -1,23 +1,41 @@
 /* SPDX-License-Identifier: GPL-2.0-or-later */
 /*
  * Copyright (c) 2021-2022 Rockchip Electronics Co., Ltd.
- * Copyright (c) 2024 Collabora Ltd.
+ * Copyright (c) 2024-2026 Collabora Ltd.
  */
 
 #ifndef __DW_HDMI_QP__
 #define __DW_HDMI_QP__
 
+#include <linux/types.h>
+
 struct device;
 struct drm_encoder;
 struct dw_hdmi_qp;
 struct platform_device;
 
+struct dw_hdmi_qp_link_cfg {
+	unsigned long long tmds_char_rate;
+	bool frl_enabled;
+	u8 frl_rate_per_lane;
+	u8 frl_lanes;
+	u8 max_frl_rate_per_lane;
+	u8 max_frl_lanes;
+	u8 min_frl_rate_per_lane;
+	u8 min_frl_lanes;
+	u8 max_ffe_level;
+	unsigned int bpc;
+};
+
 struct dw_hdmi_qp_phy_ops {
 	int (*init)(struct dw_hdmi_qp *hdmi, void *data);
 	void (*disable)(struct dw_hdmi_qp *hdmi, void *data);
 	enum drm_connector_status (*read_hpd)(struct dw_hdmi_qp *hdmi, void *data);
 	void (*enable_hpd)(struct dw_hdmi_qp *hdmi, void *data);
 	void (*disable_hpd)(struct dw_hdmi_qp *hdmi, void *data);
+	const struct dw_hdmi_qp_link_cfg *(*get_link_cfg)(struct dw_hdmi_qp *hdmi, void *data);
+	int (*set_frl_rate)(struct dw_hdmi_qp *hdmi, void *data, u8 rate_per_lane, u8 lanes);
+	int (*set_ffe_level)(struct dw_hdmi_qp *hdmi, void *data, u8 ffe_level);
 };
 
 struct dw_hdmi_qp_plat_data {
diff --git a/include/drm/display/drm_scdc.h b/include/drm/display/drm_scdc.h
index 7f0b05b2f2800..8d6d074743bf1 100644
--- a/include/drm/display/drm_scdc.h
+++ b/include/drm/display/drm_scdc.h
@@ -31,6 +31,8 @@
 #define SCDC_UPDATE_0 0x10
 #define  SCDC_RSED_UPDATE (1 << 6)
 #define  SCDC_FLT_UPDATE (1 << 5)
+#define  SCDC_FRL_START (1 << 4)
+#define  SCDC_SOURCE_TEST_UPDATE (1 << 3)
 #define  SCDC_READ_REQUEST_TEST (1 << 2)
 #define  SCDC_CED_UPDATE (1 << 1)
 #define  SCDC_STATUS_UPDATE (1 << 0)
@@ -46,11 +48,24 @@
 #define  SCDC_SCRAMBLING_STATUS (1 << 0)
 
 #define SCDC_CONFIG_0 0x30
+#define  SCDC_FLT_NO_RETRAIN (1 << 1)
 #define  SCDC_READ_REQUEST_ENABLE (1 << 0)
 
 #define SCDC_CONFIG_1 0x31
-#define  SCDC_FRL_RATE   0x0f
-#define  SCDC_FFE_LEVELS 0xf0
+#define  SCDC_FFE_LEVELS_MASK 0xf0
+#define  SCDC_FFE_LEVELS SCDC_FFE_LEVELS_MASK
+#define  SCDC_FRL_RATE_MASK 0x0f
+#define  SCDC_FRL_RATE SCDC_FRL_RATE_MASK
+#define  SCDC_FRL_RATE_DISABLE 0
+#define  SCDC_FRL_RATE_3GBPS_3LANE 1
+#define  SCDC_FRL_RATE_6GBPS_3LANE 2
+#define  SCDC_FRL_RATE_6GBPS_4LANE 3
+#define  SCDC_FRL_RATE_8GBPS_4LANE 4
+#define  SCDC_FRL_RATE_10GBPS_4LANE 5
+#define  SCDC_FRL_RATE_12GBPS_4LANE 6
+
+#define SCDC_SOURCE_TEST_CONFIG 0x35
+#define  SCDC_FLT_NO_TIMEOUT (1 << 5)
 
 #define SCDC_STATUS_FLAGS_0 0x40
 #define  SCDC_DSC_FAIL (1 << 7)
@@ -64,9 +79,13 @@
 #define  SCDC_CLOCK_DETECT (1 << 0)
 
 #define SCDC_STATUS_FLAGS_1 0x41
+#define  SCDC_FRL_LN1_LTP_REQ_MASK 0xf0
+#define  SCDC_FRL_LN0_LTP_REQ_MASK 0x0f
 #define  SCDC_LN_EVEN_TRAIN_PTRN 0x0f
 #define  SCDC_LN_ODD_TRAIN_PTRN 0xf0
 #define SCDC_STATUS_FLAGS_2 0x42
+#define  SCDC_FRL_LN3_LTP_REQ_MASK 0xf0
+#define  SCDC_FRL_LN2_LTP_REQ_MASK 0x0f
 
 #define SCDC_ERR_DET_0_L 0x50
 #define SCDC_ERR_DET_0_H 0x51
diff --git a/include/drm/display/drm_scdc_helper.h b/include/drm/display/drm_scdc_helper.h
index 7ce4b8af893d4..171c8d4e57aed 100644
--- a/include/drm/display/drm_scdc_helper.h
+++ b/include/drm/display/drm_scdc_helper.h
@@ -173,4 +173,10 @@ int drm_scdc_read_state(struct drm_connector *connector,
 			struct drm_scdc_state *state);
 void drm_scdc_debugfs_init(struct drm_connector *connector, struct dentry *root);
 
+bool drm_scdc_set_frl(struct drm_connector *connector,
+		      u8 rate_per_lane, u8 lanes, u8 max_ffe_level);
+bool drm_scdc_calc_lower_frl(u8 in_rate_per_lane, u8 in_lanes,
+			     u8 *out_rate_per_lane, u8 *out_lanes);
+bool drm_scdc_get_frl_ltp_request(struct drm_connector *connector,
+				  u8 *ln0, u8 *ln1, u8 *ln2, u8 *ln3);
 #endif
diff --git a/include/drm/drm_bridge.h b/include/drm/drm_bridge.h
index ef402a875862e..7761660a0e6a5 100644
--- a/include/drm/drm_bridge.h
+++ b/include/drm/drm_bridge.h
@@ -535,10 +535,12 @@ struct drm_bridge_funcs {
 	 *
 	 * Check if anything is attached to the bridge output.
 	 *
-	 * This callback is optional, if not implemented the bridge will be
-	 * considered as always having a component attached to its output.
-	 * Bridges that implement this callback shall set the
-	 * DRM_BRIDGE_OP_DETECT flag in their &drm_bridge->ops.
+	 * This is the non-atomic version of detect_ctx() callback, and is
+	 * optional. If both are implemented, it is ignored. If none is
+	 * implemented, the bridge will be considered as always having a
+	 * component attached to its output. Bridges that implement this
+	 * callback shall set the DRM_BRIDGE_OP_DETECT flag in their
+	 * &drm_bridge->ops.
 	 *
 	 * RETURNS:
 	 *
@@ -547,6 +549,32 @@ struct drm_bridge_funcs {
 	enum drm_connector_status (*detect)(struct drm_bridge *bridge,
 					    struct drm_connector *connector);
 
+	/**
+	 * @detect_ctx:
+	 *
+	 * Check if anything is attached to the bridge output.
+	 *
+	 * This is the atomic version of detect() callback, and is optional.
+	 * If both are implemented, it takes precedence. If none is implemented,
+	 * the bridge will be considered as always having a component attached
+	 * to its output. Bridges that implement this callback shall set the
+	 * DRM_BRIDGE_OP_DETECT flag in their &drm_bridge->ops.
+	 *
+	 * To avoid races against concurrent connector state updates, the
+	 * helper libraries always call this with ctx set to a valid context,
+	 * and &drm_mode_config.connection_mutex will always be locked with
+	 * the ctx parameter set to this ctx. This allows taking additional
+	 * locks as required.
+	 *
+	 * RETURNS:
+	 *
+	 * &drm_connector_status indicating the bridge output status,
+	 * or the error code returned by drm_modeset_lock(), -EDEADLK.
+	 */
+	int (*detect_ctx)(struct drm_bridge *bridge,
+			  struct drm_connector *connector,
+			  struct drm_modeset_acquire_ctx *ctx);
+
 	/**
 	 * @get_modes:
 	 *
@@ -716,6 +744,29 @@ struct drm_bridge_funcs {
 	 */
 	int (*hdmi_scrambler_disable)(struct drm_bridge *bridge);
 
+	/**
+	 * @hdmi_frl_rate_valid:
+	 *
+	 * Check whether a particular FRL data rate (given in bps) is
+	 * supported by the driver by using a FRL link rate that doesn't
+	 * exceed the maximum advertised by the sink (also given in bps).
+	 *
+	 * This callback is optional and should only be implemented by the
+	 * bridges that take part in the HDMI connector implementation and
+	 * need to advertise FRL support.  Bridges that implement it shall
+	 * set the DRM_BRIDGE_OP_HDMI flag in their &drm_bridge->ops.
+	 *
+	 * Returns:
+	 *
+	 * Either &drm_mode_status.MODE_OK or one of the failure reasons
+	 * in &enum drm_mode_status.
+	 */
+	enum drm_mode_status
+	(*hdmi_frl_rate_valid)(const struct drm_bridge *bridge,
+			       const struct drm_display_mode *mode,
+			       unsigned long long frl_data_rate,
+			       unsigned long long frl_max_link_rate);
+
 	/**
 	 * @hdmi_clear_avi_infoframe:
 	 *
@@ -1051,8 +1102,8 @@ struct drm_bridge_timings {
 enum drm_bridge_ops {
 	/**
 	 * @DRM_BRIDGE_OP_DETECT: The bridge can detect displays connected to
-	 * its output. Bridges that set this flag shall implement the
-	 * &drm_bridge_funcs->detect callback.
+	 * its output. Bridges that set this flag shall implement either the
+	 * &drm_bridge_funcs->detect or &drm_bridge_funcs->detect_ctx callbacks.
 	 */
 	DRM_BRIDGE_OP_DETECT = BIT(0),
 	/**
@@ -1635,6 +1686,9 @@ drm_atomic_helper_bridge_propagate_bus_fmt(struct drm_bridge *bridge,
 
 enum drm_connector_status
 drm_bridge_detect(struct drm_bridge *bridge, struct drm_connector *connector);
+int drm_bridge_detect_ctx(struct drm_bridge *bridge,
+			  struct drm_connector *connector,
+			  struct drm_modeset_acquire_ctx *ctx);
 int drm_bridge_get_modes(struct drm_bridge *bridge,
 			 struct drm_connector *connector);
 const struct drm_edid *drm_bridge_edid_read(struct drm_bridge *bridge,
diff --git a/include/drm/drm_connector.h b/include/drm/drm_connector.h
index 09be9288b2c08..8fb3109eab84c 100644
--- a/include/drm/drm_connector.h
+++ b/include/drm/drm_connector.h
@@ -1474,6 +1474,28 @@ struct drm_connector_hdmi_funcs {
 				const struct drm_display_mode *mode,
 				unsigned long long tmds_rate);
 
+	/**
+	 * @frl_rate_valid:
+	 *
+	 * This callback is invoked at atomic_check time to figure out
+	 * whether a particular FRL data rate (given in bps) is supported
+	 * by the driver by using a FRL link rate that doesn't exceed the
+	 * maximum advertised by the sink (also expressed in bps).
+	 *
+	 * The @frl_rate_valid callback is only mandatory when drivers
+	 * need to advertise FRL support.
+	 *
+	 * Returns:
+	 *
+	 * Either &drm_mode_status.MODE_OK or one of the failure reasons
+	 * in &enum drm_mode_status.
+	 */
+	enum drm_mode_status
+	(*frl_rate_valid)(const struct drm_connector *connector,
+			  const struct drm_display_mode *mode,
+			  unsigned long long frl_data_rate,
+			  unsigned long long frl_max_link_rate);
+
 	/**
 	 * @read_edid:
 	 *
diff --git a/include/linux/phy/phy-hdmi.h b/include/linux/phy/phy-hdmi.h
index d4cf4430ee8f3..1d4b624750797 100644
--- a/include/linux/phy/phy-hdmi.h
+++ b/include/linux/phy/phy-hdmi.h
@@ -19,6 +19,10 @@ enum phy_hdmi_mode {
  * @tmds_char_rate: HDMI TMDS Character Rate in Hertz.
  * @frl.rate_per_lane: HDMI FRL Rate per Lane in Gbps.
  * @frl.lanes: HDMI FRL lanes count.
+ * @frl.ffe_level: Transmitter Feed Forward Equalizer Level.
+ *	Optional, only meaningful when set_ffe_level flag is on.
+ * @frl.set_ffe_level: Flag indicating whether or not to reconfigure ffe_level.
+ *	All the other struct fields must be ignored when this is used.
  *
  * This structure is used to represent the configuration state of a HDMI phy.
  */
@@ -29,6 +33,8 @@ struct phy_configure_opts_hdmi {
 		struct {
 			u8 rate_per_lane;
 			u8 lanes;
+			u8 ffe_level;
+			u8 set_ffe_level : 1;
 		} frl;
 	};
 };
