From 75dbc7c401b2408c7ae46c4acb1be808e79664e1 Mon Sep 17 00:00:00 2001
From: Yi Ding <yi.s.ding@gmail.com>
Date: Tue, 14 Jul 2026 20:56:03 -0700
Subject: [PATCH] daemon-handover: Release taken socket connection

---
 src/grd-daemon-handover.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/grd-daemon-handover.c b/src/grd-daemon-handover.c
index 709f14b..cf82509 100644
--- a/src/grd-daemon-handover.c
+++ b/src/grd-daemon-handover.c
@@ -93,7 +93,7 @@ on_take_client_finished (GObject      *object,
   GrdDBusRemoteDesktopRdpHandover *proxy;
   GrdDaemonHandover *daemon_handover;
   GrdRdpServer *rdp_server;
-  GSocketConnection *socket_connection;
+  g_autoptr (GSocketConnection) socket_connection = NULL;
 
   proxy = GRD_DBUS_REMOTE_DESKTOP_RDP_HANDOVER (object);
   if (!grd_dbus_remote_desktop_rdp_handover_call_take_client_finish (
-- 
2.55.0


